wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt

```diff
@@ -0,0 +1,276 @@
+  }
+
+  internal companion object {
+    internal fun <S : Any, R : Any> Method.toGrpc(): GrpcMethod<S, R> {
+      val wireRpc = getAnnotation(WireRpc::class.java)
+      val requestAdapter = ProtoAdapter.get(wireRpc.requestAdapter) as ProtoAdapter<S>
+      val responseAdapter = ProtoAdapter.get(wireRpc.responseAdapter) as ProtoAdapter<R>
+      val parameterTypes = genericParameterTypes
+      val returnType = genericReturnType
+
+      if (parameterTypes.size == 2) {
+        // Coroutines request-response.
+        // Object methodName(RequestType, Continuation)
+        if (parameterTypes[1].rawType() == Continuation::class.java) {
+          return RequestResponse(wireRpc.path, requestAdapter, responseAdapter)
+        }
+
+      } else if (parameterTypes.size == 1) {
+        // Coroutines streaming response.
+        // ReceiveChannel<ResponseType> methodName(RequestType)
+        if (returnType.rawType() == ReceiveChannel::class.java) {
+          return StreamingResponse(wireRpc.path, requestAdapter, responseAdapter)
+        }
+
+        // Blocking streaming response.
+        // MessageSource<ResponseType> methodName(RequestType)
+        if (returnType.rawType() == MessageSource::class.java) {
+          return BlockingStreamingResponse(wireRpc.path, requestAdapter, responseAdapter)
+        }
+
+        // Blocking request-response.
+        // ResponseType methodName(RequestType)
+        return BlockingRequestResponse(wireRpc.path, requestAdapter, responseAdapter)
+
+      } else if (parameterTypes.isEmpty()) {
+        if (returnType.rawType() == Pair::class.java) {
+          val pairType = returnType as ParameterizedType
+          val requestType = pairType.genericParameterType(index = 0).rawType()
+          val responseType = pairType.genericParameterType(index = 1).rawType()
+
+          // Coroutines full duplex.
+          // Pair<SendChannel<RequestType>, ReceiveChannel<ResponseType>> methodName()
+          if (requestType == SendChannel::class.java
+              && responseType == ReceiveChannel::class.java) {
+            return FullDuplex(wireRpc.path, requestAdapter, responseAdapter)
+          }
+
+          // Coroutines streaming request.
+          // Pair<SendChannel<RequestType>, Deferred<ResponseType>> methodName()
+          if (requestType == SendChannel::class.java
+              && responseType == Deferred::class.java) {
+            return StreamingRequest(wireRpc.path, requestAdapter, responseAdapter)
+          }
+
+          // Blocking full duplex OR streaming request. (single response could be Future instead?)
+          // Pair<MessageSink<RequestType>, MessageSource<ResponseType>> methodName()
+          if (requestType == MessageSink::class.java
+              && responseType == MessageSource::class.java) {
+            return BlockingFullDuplex(wireRpc.path, requestAdapter, responseAdapter)
+          }
+        }
+      }
+
+      error("unexpected gRPC method: $this")
+    }
+  }
+}
```

Comment:
Check that IDE box!

Reply:
![Screen Shot 2019-08-16 at 11 04 33 AM](https://user-images.githubusercontent.com/133019/63177399-cdbbe480-c015-11e9-8f22-bfe37a0bb573.png)

Reply:
Added.

Reply:
Also, invisible whitespace is dumb and I’m annoyed that this is even a warning that exists.

Reply:
I don't disagree. Trailing newlines are nice because when you add new code to the end of a file you see only added lines instead of a changed line and then added lines. Not sure why it's still opt-in.